snuba/utils/bucket_timer.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+from collections import defaultdict, deque
+from datetime import datetime, timedelta
+from typing import Deque, List, Optional
+
+from snuba import environment
+from snuba.utils.metrics.wrapper import MetricsWrapper
+
+WINDOW_SIZE = timedelta(minutes=10)  # minutes
+
+
+class Bucket:
+    def __init__(
+        self,
+        project_id: Optional[int],
+        minute: datetime,
+        processing_time: timedelta = timedelta(seconds=0),
+    ) -> None:
+        self.project_id = project_id  # is None for global bucket
+        self.minute = minute
+        self.processing_time = processing_time
+
+    def add_processing_time_seconds(self, seconds: timedelta) -> None:
+        self.processing_time += seconds
+
+
+class Counter:
+    def __init__(self, consumer_group: str) -> None:
+        self.consumer_group: str = consumer_group
+        self.buckets: Deque[Bucket] = deque()
+
+    def get_time_sum(self, group: List[timedelta]) -> timedelta:
+        return sum([processing_time for processing_time in group], timedelta())
+
+    def floor_minute(self, time: datetime) -> datetime:
+        return time - timedelta(seconds=time.second, microseconds=time.microsecond)
+
+    def ceil_minute(self, time: datetime) -> datetime:
+        if time.second == 0 and time.microsecond == 0:
+            return time
+        return self.floor_minute(time + timedelta(minutes=1))
+
+    def trim_expired_buckets(self, now: datetime) -> None:
+        current_minute = self.floor_minute(now)
+        window = current_minute - WINDOW_SIZE
+        while self.buckets and self.buckets[0].minute < window:
+            self.buckets.popleft()
+
+    def get_existing_bucket(
+        self, project_id: Optional[int], minute: datetime
+    ) -> Optional[Bucket]:
+        for bucket in self.buckets:
+            if bucket.project_id == project_id and bucket.minute == minute:
+                return bucket
+        return None
+
+    def create_and_add_bucket(
+        self,
+        project_id: Optional[int],
+        start_minute: datetime,
+        processing_time: timedelta,
+    ) -> None:
+        bucket = self.get_existing_bucket(project_id, start_minute)
+        if not bucket:
+            bucket = Bucket(project_id, start_minute, processing_time)
+            self.buckets.append(bucket)
+        else:
+            bucket.add_processing_time_seconds(processing_time)
+
+    def write_to_bucket(
```

Comment:
nit: naming. The public interface of a class should be intuitive and give you an idea of what the class does for you (more interesting that how it does it). A good interface is one you do not know how to document because it is obvious. Think about a list: `append`, `get`. 

If you were describing what this class does you would probably explain it as:
- it accounts for the time spent on some activity for a project
- it returns the project ids that spent too much time.

So I would call this method something like `account`, `record_time_spent`. And I would call `compare_and_write_metric` something like `get_top_projects` or `get_greedy_projects`, or anything that give you an idea of a property of the projects you are returning. 

Calling this method `write_to_bucket` does not tell me what you are writing and why I would use it. On the other end it exposes an implementation detail (buckets) that the call site should not know about because it is irrelevant to the interface.
Calling the method below `compare_and_write_metric` instead does not tell me what it is comparing with what. 
Basically this naming would assume the call site knows the internals of the class, which is exactly what you want to hide by adding a class.